### ISSUE TITLE
Spec should say when 'local' files are overwritten (5.1)

### ISSUE THREAD

**Issue by edsko** [theupdateframework/tuf#283]
(I'm logging issues such as these mostly to help make the spec more precise. I hope it's useful.)

The update process described in Section 5.1 looks something like:

1. Download timestamp, verify, and compare to old timestamp. 
2. If changed, download snapshot, verify, and compare to old snapshot.
3. If root metadata changed, download new, verify, and start over.
4. Download and verify other changed top-level metadata.

What does this _not_ specify is when, say, the "old" timestamp that the client cached locally is _replaced_ with the new timestamp from the server. This is somewhat subtle; it is not okay to replace the local copy of the timestamp with the server copy as soon as it has been verified and compared, because then we might do:

1. Download timestamp, verify, compare to local. Notice changed, replace local copy.
2. Download snapshot, verify, compare to local, notice root changed (perhaps just a new root key was added, nothing else), replace local copy.
3. Download root, verify, start process over.
4. Now we download the timestamp once again, verify it, and compare it to our local copy. _Since we already replaced the local copy, we conclude it's not changed_ and we return "No changes" to the client.

This means we might fail to notice that there are other top-level metadata files that ought to have been replaced. 

Incidentally, on a verification _error_ in our implementation we actually _delete_ the local timestamp and local snapshot after getting (and verifying) the new root info. We do this because, say, the timestamp key has been compromised and the attacker has set the file version to MAX_INT. Then if we don't do something to deal with this situation, clients would subsequently reject all further updates because the file version cannot change anymore. By removing the local files we basically reset the clients state to "we don't know what's on the server".

**Comment by vladimir-v-diaz** [comment]
"Now we download the timestamp once again, verify it, and compare it to our local copy. Since we already replaced the local copy, we conclude it's not changed and we return "No changes" to the client."

I was confused about the "No changes" message being returned to the client, but then I noticed that the [spec does mention it](https://github.com/theupdateframework/tuf/blob/develop/docs/tuf-spec.txt#L924-L926).  This is a mistake.  The method that refreshes the top-level metadata should always download the timestamp role, and check that the the other top-level are the latest.

It's okay for the updater to log useful information about the update call (e.g., the timestamp is up-to-date), but error messages should normally only occur for things like an expired root file, or attacks that are detected.

We'll fix this in the updated TUF specification.

**Comment by edsko** [comment]
Now I'm confused :) The "no changes" message is no error message -- when the check for updates notices that the timestamp is still the same, then surely it is okay for it to conclude that there are, in fact, no changes, and that none of the other top-level metadata files need to be checked?

**Comment by vladimir-v-diaz** [comment]
The current spec states:
"If the timestamp file lists the same snapshot.json file as was previously seen, the client code is informed that no updates are available and **the update checking process stops.**"

I should have emphasized that prematurely stopping the update process is the mistake IMO.  I'm not sure why I mentioned error messages in the third paragraph, but yes, the update routine should not return control to the client code if the timestamp indicates that there are no changes.  It should continue on with the other top-level roles to ensure that the client possesses all the required metadata, and that they are up-to-date.

It may be the case that the client initially has only the root of trust (`root.json`), or that `targets.json` has expired or is outdated.  For example,

`root.json` is locally cached and outdated.

The update routine would do:
1.  Download `timestamp.json`.
2.  Download `snapshot.json`, if changed.
3.  Replace `root.json` if locally cached version is not latest, and go back to step 1.
4.  Download `targets.json`, if changed.

If the updater stops after detecting that timestamp is up-to-date (on the second iteration of the routine above), then the client is never able to retrieve the latest copy of `targets.json` in step 4.

So, it is okay to log that the timestamp is up-to-date, but don't return control to the client code.

**Comment by vladimir-v-diaz** [comment]
>We do this because, say, the timestamp key has been compromised and the attacker has set the file version to MAX_INT. Then if we don't do something to deal with this situation, clients would subsequently reject all further updates because the file version cannot change anymore. By removing the local files we basically reset the clients state to "we don't know what's on the server".

I agree that clients should subsequently reject all further updates if valid top-level metadata cannot be fetched.  I'm not sure if removing the local files is necessary, though.  I say this because the updater would always make an attempt to get the latest top-level metadata.  `root.json` should definitely not be removed, even if it has expired, since the updater uses the currently trusted root keys to allow outdated clients to update `root.json`.

If after re-downloading `root.json` valid metadata cannot be obtained for `timestamp.json`, then the updater should return an error so that clients are aware that a freeze attack might be the cause.

**Comment by edsko** [comment]
Our reasoning was as follows: suppose the snapshot key is compromised, and an attacker uses this to set the version on the `snapshot.json` to be `MAX_INT` (so that it cannot be incremented anymore). Then how will clients recover from this situation? At some point server admins discover that the key is compromised and update the `root.json` to record a new snapshot key. But unless clients deal with this situation specifically, they will still not be able to download a new snapshot because the version will still not be newer than the version they previously downloaded. That's why we delete the local snapshot/timestamp when the root info is updated: it basically resets the versioning scheme. But perhaps there is a better way to do it?

**Comment by edsko** [comment]
> The current spec states
> "If the timestamp file lists the same snapshot.json file as was previously seen, the client code is informed that no updates are available and the update checking process stops."

> I should have emphasized that prematurely stopping the update process is the mistake IMO. I'm not sure why I mentioned error messages in the third paragraph, but yes, the update routine should not return control to the client code if the timestamp indicates that there are no changes. It should continue on with the other top-level roles to ensure that the client possesses all the required metadata, and that they are up-to-date.

I don't understand this. All top-level metadata are recorded with their hash in the snapshot, so if they change, then the snapshot changes. And since the hash in the snapshot is recorded in the timestamp, this means that the timestamp changes. In other words, if the timestamp hasn't changed, no other file can have changed. 

(In our implementation, "to have changed compared to the local copy" is interpreted as "if the client does not yet have a local copy, regard as changed".)

**Comment by vladimir-v-diaz** [comment]
>> That's why we delete the local snapshot/timestamp when the root info is updated: it basically resets the versioning scheme. But perhaps there is a better way to do it?

It makes sense to replace snapshot/timestamp if `root.json` is replaced.  If a key is replaced (perhaps because the repository was compromised), it might even be safest if all local metadata is reset (according to the new root.json).  We'll add this to the spec.

>> I don't understand this. All top-level metadata are recorded with their hash in the snapshot, so if they change, then the snapshot changes. And since the hash in the snapshot is recorded in the timestamp, this means that the timestamp changes. In other words, if the timestamp hasn't changed, no other file can have changed.

>> (In our implementation, "to have changed compared to the local copy" is interpreted as "if the client does not yet have a local copy, regard as changed".)

This approach also works.  The only other issue to account for would be expired top-level metadata that isn't checked if timestamp is found to be up-to-date.

**Comment by edsko** [comment]
Just to be clear (it was probably clear to you, but it wasn't necessarily clear to me):

> It makes sense to replace snapshot/timestamp if root.json is replaced.

This should _only_ happen if the root info is _actually_ updated. If we download the root info because of a validation error, but find that the root info hasn't actually changed, we should not delete these local cached files. Otherwise we get this scenario:

1. Version in the timestamp is wrong
2. Client downloads the timestamp, compares version against cached timestamp, finds it is wrong.
3. Downloads root info (unchanged), deletes timestamp (incorrectly), and tries again.
4. Client downloads timestamp, doesn't verify version, because it doesn't have a local copy yet.

This would render file versions completely useless.
